### PR TITLE
Allow inverse scaling selection in skin editor

### PR DESCRIPTION
**Pull request by Wleter**
Addresses #13398, also slightly changed the way rotated selections scale that in my opinion feel better than earlier.

I'm not sure about the FlipScaleHandles methods, however I didn't see better way to do this.

https://github.com/ppy/osu/assets/34943251/35e72625-03c1-4f7c-8d1e-92a964112e3f

### CODE REVIEW

**Comment by peppy** [comment]
This definitely looks like how I'd expect this to work visually, so nice work on that.

I haven't checked the code yet.

**Review by peppy — DISMISSED** [review]
I think this looks pretty good

**Comment by peppy** [comment]
@bdach going to wait on your approval for this since you mentioned seeing a potential issue.

File: osu.Game/Screens/Edit/Compose/Components/SelectionBox.cs
Comment (bdach) [inline osu.Game/Screens/Edit/Compose/Components/SelectionBox.cs:310]: I'm pretty vehemently against this existing in the current shape. It's pretty much a direct violation of any sort of encapsulation. It's a method made to be called in this one specific scenario.

A way to get rid of this and preserve the behaviour is something along the lines of

```diff
diff --git a/osu.Game/Overlays/SkinEditor/SkinSelectionHandler.cs b/osu.Game/Overlays/SkinEditor/SkinSelectionHandler.cs
index ff53095e22..b30351f61b 100644
--- a/osu.Game/Overlays/SkinEditor/SkinSelectionHandler.cs
+++ b/osu.Game/Overlays/SkinEditor/SkinSelectionHandler.cs
@@ -81,18 +81,12 @@ public override bool HandleScale(Vector2 scale, Anchor anchor)
 
             if (adjustedRect.Width <= 0 || adjustedRect.Height <= 0)
             {
-                if (adjustedRect.Width <= 0)
-                {
-                    SelectionBox.FlipScaleHandles(Direction.Horizontal);
-                    HandleFlip(Direction.Horizontal, false);
-                }
+                Axes toFlip = Axes.None;
 
-                if (adjustedRect.Height <= 0)
-                {
-                    SelectionBox.FlipScaleHandles(Direction.Vertical);
-                    HandleFlip(Direction.Vertical, false);
-                }
+                if (adjustedRect.Width <= 0) toFlip |= Axes.X;
+                if (adjustedRect.Height <= 0) toFlip |= Axes.Y;
 
+                SelectionBox.PerformFlipFromScaleHandles(toFlip);
                 return true;
             }
 
diff --git a/osu.Game/Screens/Edit/Compose/Components/SelectionBox.cs b/osu.Game/Screens/Edit/Compose/Components/SelectionBox.cs
index bbf9ea8c3c..0c19f6c62e 100644
--- a/osu.Game/Screens/Edit/Compose/Components/SelectionBox.cs
+++ b/osu.Game/Screens/Edit/Compose/Components/SelectionBox.cs
@@ -4,6 +4,7 @@
 using System;
 using osu.Framework.Allocation;
 using osu.Framework.Bindables;
+using osu.Framework.Extensions.EnumExtensions;
 using osu.Framework.Graphics;
 using osu.Framework.Graphics.Containers;
 using osu.Framework.Graphics.Shapes;
@@ -307,7 +308,24 @@ private SelectionBoxButton addButton(IconUsage icon, string tooltip, Action acti
             return button;
         }
 
-        public void FlipScaleHandles(Direction direction) => dragHandles.FlipScaleHandles(direction);
+        /// <remarks>
+        /// This method should be called when a selection needs to be flipped
+        /// because of an ongoing scale handle drag that would otherwise cause width or height to go negative.
+        /// </remarks>
+        public void PerformFlipFromScaleHandles(Axes axes)
+        {
+            if (axes.HasFlagFast(Axes.X))
+            {
+                dragHandles.FlipScaleHandles(Direction.Horizontal);
+                OnFlip?.Invoke(Direction.Horizontal, false);
+            }
+
+            if (axes.HasFlagFast(Axes.Y))
+            {
+                dragHandles.FlipScaleHandles(Direction.Vertical);
+                OnFlip?.Invoke(Direction.Vertical, false);
+            }
+        }
 
         private void addScaleHandle(Anchor anchor)
         {

```

It's only _slightly_ less bad but it's not an outright "go mess in the guts of this component 2 layers down in a way that makes sense nowhere else". It has a _semblance_ of a well-defined operation.

Ideally the drag handle would be able to handle this entirely itself but that's too involved as things stand. Maybe once I get around to refactoring these at some point the same way I did rotation.

For the time being - any objections to the above?

**Review by bdach — CHANGES_REQUESTED** [review]

File: osu.Game/Screens/Edit/Compose/Components/SelectionBox.cs
Comment (Wleter) [reply osu.Game/Screens/Edit/Compose/Components/SelectionBox.cs:310]: No objections, I also wasn't sure about that one, as I said at the end of the PR message.

**Review by bdach — APPROVED** [review]

**Pull request merged**
